libpairinteraction/SystemBase.h

```diff
@@ -516,35 +521,128 @@ class SystemBase {
+        mkl_set_interface_layer(MKL_INTERFACE_LP64);
+
+        // Estimate number of found eigenvalues
+        std::vector<double> diagonal_max(hamiltonian.outerSize(), 0);
+        std::vector<double> diagonal_val;
+        std::vector<int> diagonal_idx;
+        diagonal_val.reserve(hamiltonian.outerSize());
+        diagonal_idx.reserve(hamiltonian.outerSize());
+        for (int k = 0; k < hamiltonian.outerSize(); ++k) {
+            double val = 0;
+            for (eigen_iterator_t triple(hamiltonian, k); triple; ++triple) {
+                if (triple.row() == triple.col()) {
+                    val = std::real(triple.value());
+                } else if (triple.row() != triple.col()) {
+                    diagonal_max[k] = std::max(diagonal_max[k], std::abs(triple.value()));
+                }
+            }
+            diagonal_idx.push_back(k);
+            diagonal_val.push_back(val);
+        }
+
+        MKL_INT m0 =
+            std::count_if(diagonal_val.begin(), diagonal_val.end(),
+                          [&energy_lower_bound, &energy_upper_bound](const double &val) {
+                              return (val < energy_upper_bound) && (val > energy_lower_bound);
+                          }) +
+            std::count_if(diagonal_idx.begin(), diagonal_idx.end(),
+                          [&energy_lower_bound, &energy_upper_bound, &diagonal_val,
+                           &diagonal_max](const int &i) {
+                              return ((diagonal_val[i] >= energy_upper_bound) ||
+                                      (diagonal_val[i] <= energy_lower_bound)) &&
+                                  (diagonal_val[i] < energy_upper_bound + diagonal_max[i]) &&
+                                  (diagonal_val[i] > energy_lower_bound - diagonal_max[i]);
+                          });
+
+        // Inplace conversion of the Hamiltonian to CSR with one-based indexing
+        hamiltonian = hamiltonian.transpose();
+        hamiltonian.makeCompressed();
+        for (int i = 0; i < hamiltonian.rows() + 1; ++i) {
+            hamiltonian.outerIndexPtr()[i] += 1;
+        }
+        for (int i = 0; i < hamiltonian.nonZeros(); ++i) {
+            hamiltonian.innerIndexPtr()[i] += 1;
+        }
 
-        // Get eigenvalues and eigenvectors
-        eigen_vector_double_t evals = eigensolver.eigenvalues();
-        eigen_sparse_t evecs = eigensolver.eigenvectors().sparseView();
+        // Set default parameters for the diagonalization as described at
+        // https://software.intel.com/en-us/mkl-developer-reference-c-extended-eigensolver-input-parameters.
+        // Note that fpm[0]=1 enables terminal output and fpm[26]=1 enables a matrix checker.
+
+        MKL_INT fpm[128];
+        feastinit(fpm);
+        fpm[0] = 1;
+        fpm[1] = 6;
+        fpm[26] = 0;
+        fpm[3] = 5;
+        if (threshold != std::numeric_limits<double>::max()) {
+            fpm[2] = std::min(std::round(-std::log10(threshold)),
```

Comment:
This will blow up if a user sets `threshold = 0`.  Maybe use
```cpp
std::log10(1 + threshold) // potential problem for -1
```
or
```cpp
std::log10(threshold == 0 ? 1 : threshold) // extra branch
```